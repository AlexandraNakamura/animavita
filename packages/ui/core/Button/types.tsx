import {TouchableOpacityProps} from 'react-native';
import {DefaultTheme} from 'styled-components/native';

// gradient
export const gradientConfig = {
  start: {
    x: 0,
    y: 0,
  },
  end: {
    x: 1,
    y: 1,
  },
  locations: [0.1, 0.9],
};

// sizes
export const LARGE = 'large';
export const SMALL = 'small';

export type ButtonActive = boolean;
export type ButtonSize = typeof LARGE | typeof SMALL;
export type ButtonRounded = boolean;
export type ButtonText = string;
export type ButtonDisabled = boolean;
export type ButtonOutline = boolean;
export type ButtonTextColor = string;

export interface ButtonProps extends TouchableOpacityProps {
  readonly rounded?: ButtonRounded;
  readonly active?: ButtonActive;
  readonly gradient?: ButtonActive;
  readonly outline?: ButtonOutline;
  readonly textColor?: ButtonTextColor;
  disabled?: ButtonDisabled;
  text: ButtonText;
  size: ButtonSize;
}

export type ButtonComponentType = React.FC<ButtonProps & {theme: DefaultTheme}>;
