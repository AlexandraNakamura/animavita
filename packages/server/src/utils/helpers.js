import AWS from 'aws-sdk';

require('dotenv').config();

const {
  S3_KEY, S3_SECRET, S3_BUCKET, S3_REGION
} = process.env;

export const getAnimalGender = (gender) => {
  const genders = [
    {
      key: 'male',
      translate: 'Macho'
    },
    {
      key: 'female',
      translate: 'Fêmea'
    }
  ];

  return genders.find(element => element.key === gender).translate;
};

export const getAnimalSize = (size) => {
  const sizes = [
    {
      key: 'small',
      translate: 'Pequeno'
    },
    {
      key: 'medium',
      translate: 'Médio'
    },
    {
      key: 'bigger',
      translate: 'Grande'
    }
  ];

  return sizes.find(element => element.key === size).translate;
};

export const uploadFile = async (upload) => {
  let url = '';
  // eslint-disable-next-line new-cap
  const base64Data = new Buffer.from(
    upload.preview.uri.replace(/^data:image\/\w+;base64,/, ''),
    'base64'
  );

  const params = {
    Bucket: S3_BUCKET,
    Key: upload.image.name,
    Body: base64Data,
    ACL: 'public-read',
    ContentEncoding: 'base64',
    ContentType: upload.image.type
  };

  const s3 = new AWS.S3({
    accessKeyId: S3_KEY,
    secretAccessKey: S3_SECRET,
    region: S3_REGION
  });

  try {
    const { Location } = await s3.upload(params).promise();
    url = Location;
  } catch (err) {
    throw err;
  }

  return url;
};
