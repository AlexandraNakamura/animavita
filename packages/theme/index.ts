export {default as StyledTheme} from './src/Theme';
export {ThemeContextProvider, withThemeProvider} from './src/Context';
export * from './src/PixelRatio';
