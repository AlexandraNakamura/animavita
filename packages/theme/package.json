{
  "name": "@animavita/theme",
  "version": "1.0.0",
  "private": true,
  "dependencies": {
    "react": "16.12.0",
    "react-native": "https://github.com/expo/react-native/archive/sdk-36.0.1.tar.gz",
    "styled-components": "5.0.0"
  },
  "devDependencies": {
    "@types/styled-components": "^4.4.2"
  }
}
