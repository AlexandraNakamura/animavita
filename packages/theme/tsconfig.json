{
  "compilerOptions": {
    "rootDir": ".",
    "composite": true,
    "noEmit": true,
    "typeRoots": ["@types"]
  },
  "extends": "../../tsconfig.json",
  "include": ["."]
}
