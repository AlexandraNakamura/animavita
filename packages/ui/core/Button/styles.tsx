import styled, {css} from 'styled-components/native';
import {px2ddp} from '@animavita/theme';
import {LinearGradient} from 'expo-linear-gradient';

import Typography from '../Typography';

import {ButtonActive, ButtonSize, ButtonOutline, LARGE, SMALL, ButtonRounded, ButtonDisabled} from './types';

export const DisabledStyle = css``;

export const RoundedStyle = css<{size: ButtonSize}>`
  border-radius: ${({size}) => {
    switch (size) {
      case LARGE:
        return px2ddp(15);
      case SMALL:
        return px2ddp(10);
    }
  }}px;
`;

export const OutlineTypeStyle = css<{disabled?: boolean}>`
  background-color: transparent;
  border-width: ${px2ddp(0.8)}px;
  border-color: ${({theme, disabled}) => (disabled ? theme.grey : theme.greenLight)};
  ${({disabled}) => disabled && DisabledStyle}
`;

export const LargeSizeStyle = css`
  padding: ${px2ddp(6)}px ${px2ddp(12)}px;
`;

export const SmallSizeStyle = css`
  padding: ${px2ddp(4)}px ${px2ddp(10)}px;
`;

export const TouchableStyle = css<{
  active?: ButtonActive;
  gradient?: ButtonActive;
  outline?: ButtonOutline;
  rounded?: ButtonRounded;
  disabled?: ButtonDisabled;
  size: ButtonSize;
}>`
  ${({outline}) => {
    if (!outline) return;

    return OutlineTypeStyle;
  }}
  ${({rounded}) => {
    if (!rounded) return;

    return RoundedStyle;
  }}
  ${({size, active, gradient, disabled}) => {
    if (!disabled && (active || gradient)) return;

    switch (size) {
      case LARGE:
        return LargeSizeStyle;
      case SMALL:
        return SmallSizeStyle;
    }
  }}
  justify-content: center;
`;

export const Touchable = styled.TouchableOpacity<{
  active?: ButtonActive;
  gradient?: ButtonActive;
  outline?: ButtonOutline;
  rounded?: ButtonRounded;
  disabled?: ButtonDisabled;
  size: ButtonSize;
}>`
  ${TouchableStyle}
`;

export const StyledLinearGradient = styled(LinearGradient)<{size: ButtonSize}>`
  ${({size}) => {
    switch (size) {
      case LARGE:
        return LargeSizeStyle;
      case SMALL:
        return SmallSizeStyle;
    }
  }}
  border-radius: ${px2ddp(10)}px;
`;

export const Text = styled(Typography)`
  text-align: center;
`;
