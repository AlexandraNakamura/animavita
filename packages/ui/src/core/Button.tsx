import React from 'react';
import styled, {css, DefaultTheme, withTheme} from 'styled-components/native';
import {TouchableOpacityProps} from 'react-native';
import {px2ddp} from '@animavita/theme';

import Typography from './Typography';

// types
const PRIMARY = 'primary';
const OUTLINE = 'outline';
// sizes
const LARGE = 'large';
const SMALL = 'small';

const DisabledStyle = css`
  cursor: no-drop;
`;
const PrimaryTypeStyle = css<{disabled?: boolean}>`
  border-radius: ${() => `${px2ddp(10)}px`};
  ${({disabled}) =>
    disabled
      ? css`
          background-color: ${({theme}) => theme.greyLight};
          ${DisabledStyle}
        `
      : css`
          background-image: linear-gradient(
            to right,
            ${({theme}) => theme.greenDark},
            ${({theme}) => theme.greenLight}
          );
        `}
`;
const OutlineTypeStyle = css<{disabled?: boolean}>`
  background-color: transparent;
  border-radius: ${px2ddp(10)}px;
  border-width: ${px2ddp(1)}px;
  border-color: ${({theme, disabled}) => (disabled ? theme.grey : theme.greenLight)};
  ${({disabled}) => disabled && DisabledStyle}
`;
const LargeSizeStyle = css`
  padding: ${px2ddp(20)}px ${px2ddp(40)}px;
`;
const SmallSizeStyle = css`
  padding: ${px2ddp(8)}px ${px2ddp(40)}px;
`;
const TouchableStyle = css<{type: ButtonType; size: ButtonSize}>`
  ${({type}) => {
    switch (type) {
      case PRIMARY:
        return PrimaryTypeStyle;
      case OUTLINE:
        return OutlineTypeStyle;
    }
  }}
  ${({size}) => {
    switch (size) {
      case LARGE:
        return LargeSizeStyle;
      case SMALL:
        return SmallSizeStyle;
    }
  }}
`;

const Touchable = styled.TouchableOpacity<{type: ButtonType; size: ButtonSize}>`
  ${TouchableStyle}
`;
const Text = styled(Typography)`
  text-align: center;
`;

type ButtonType = typeof PRIMARY | typeof OUTLINE;
type ButtonSize = typeof LARGE | typeof SMALL;

interface ButtonProps extends TouchableOpacityProps {
  text: string;
  type: ButtonType;
  size: ButtonSize;
  disabled?: boolean;
}

type ButtonComponentType = React.FC<ButtonProps & {theme: DefaultTheme}>;

const Button: ButtonComponentType = ({text, theme, ...props}) => {
  const {type, disabled} = props;

  const primaryTextColor = disabled ? theme.grey : theme.white;
  const outlineTextColor = disabled ? theme.grey : theme.greenLight;

  const color = type === PRIMARY ? primaryTextColor : outlineTextColor;
  return (
    <Touchable {...props}>
      <Text variant="body" color={color} type="bold">
        {text}
      </Text>
    </Touchable>
  );
};

export default withTheme(Button) as React.FC<ButtonProps>;
