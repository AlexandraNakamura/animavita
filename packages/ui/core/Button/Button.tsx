import React from 'react';
import {DefaultTheme, withTheme} from 'styled-components/native';

import {Touchable, StyledLinearGradient, Text} from './styles';
import {
  ButtonComponentType,
  ButtonActive,
  ButtonDisabled,
  ButtonTextColor,
  ButtonProps,
  gradientConfig,
  SMALL as TEXT_SMALL,
} from './types';

function getColor(
  theme: DefaultTheme,
  disabled?: ButtonDisabled,
  color?: ButtonTextColor,
  active?: ButtonActive,
): string {
  if (disabled) return theme.grey;
  if (color) return color;
  if (active) return theme.white;

  return theme.greenLight;
}

const Button: ButtonComponentType = ({text, theme, children, ...props}) => {
  const {active, gradient, disabled, rounded, outline, size, textColor: color} = props;

  const textColor = getColor(theme, disabled, color, active || gradient);
  const textSize = size === TEXT_SMALL ? 'body' : 'title-3';
  const {locations, start, end} = gradientConfig;

  if (__DEV__) {
    if (disabled && (active || gradient)) console.warn("You can't use `gradient` prop when your button is disabled.");
    if (rounded && !outline && !gradient && !active)
      console.warn('To use `rounded` prop you need the `outline`, `gradient` or `active` prop as well.');
    if (outline && (active || gradient))
      console.warn("You can't use `outline` prop combined with `gradient` or `active` props.");
    if (children && text) console.warn("You can't use `title` prop combined with `children`. Use only one of both.");
  }

  return (
    <Touchable {...props}>
      {!disabled && (active || gradient) ? (
        <StyledLinearGradient
          size={size}
          locations={locations}
          start={start}
          end={end}
          colors={[theme.greenDark, theme.greenLight]}
          active={active}
          rounded={rounded}
          gradient={gradient}>
          {children ? (
            children
          ) : (
            <Text variant={textSize} color={textColor} type="bold">
              {text}
            </Text>
          )}
        </StyledLinearGradient>
      ) : children ? (
        children
      ) : (
        <Text variant={textSize} color={textColor} type="bold">
          {text}
        </Text>
      )}
    </Touchable>
  );
};

export default withTheme(Button) as React.FC<ButtonProps>;
