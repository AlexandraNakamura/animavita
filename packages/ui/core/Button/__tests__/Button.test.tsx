import React from 'react';
import {fireEvent} from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import {StyledTheme} from '@animavita/theme';

import {Mount, ThemeContext} from '../../../tests/helpers';
import Button from '../Button';

describe('Button', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const cb = jest.fn();

  it('calls onPress', () => {
    const {getByText} = Mount(<Button text="Title" size="small" onPress={cb} />);

    fireEvent.press(getByText('Title'));

    expect(cb).toHaveBeenCalled();
  });

  describe('textColor variations', () => {
    it('renders disabled color', () => {
      const {getByText} = Mount(<Button text="Title" size="small" disabled onPress={cb} />);

      expect(getByText('Title')).toHaveStyle({color: StyledTheme.grey});
    });

    it('renders custom color', () => {
      const {getByText} = Mount(<Button text="Title" size="small" textColor="red" onPress={cb} />);

      expect(getByText('Title')).toHaveStyle({color: 'red'});
    });

    it('renders active/gradient color', () => {
      const {getByText} = Mount(<Button text="Title" size="small" gradient onPress={cb} />);

      expect(getByText('Title')).toHaveStyle({color: StyledTheme.white});
    });
  });

  describe('textSize variations', () => {
    it('renders body font size', () => {
      const {getByText, debug} = Mount(<Button text="Title" size="small" disabled onPress={cb} />);

      expect(getByText('Title')).toHaveStyle({fontSize: parseFloat(StyledTheme.sizeBody)});
    });

    it('renders title-2 font size', () => {
      const {getByText, debug} = Mount(<Button text="Title" size="large" disabled onPress={cb} />);

      expect(getByText('Title')).toHaveStyle({fontSize: parseFloat(StyledTheme.sizeTitle3)});
    });
  });

  describe('when violate the constraints', () => {
    beforeEach(() => {
      global.console = {warn: jest.fn()};
    });

    it('logs disabled and gradient warn', () => {
      Mount(<Button text="Title" disabled gradient size="small" onPress={cb} />);

      expect(console.warn).toHaveBeenCalledWith("You can't use `gradient` prop when your button is disabled.");
    });

    it('logs rounded warn', () => {
      Mount(<Button text="Title" rounded size="small" onPress={cb} />);

      expect(console.warn).toHaveBeenCalledWith(
        'To use `rounded` prop you need the `outline`, `gradient` or `active` prop as well.',
      );
    });

    it('logs outline warn', () => {
      Mount(<Button text="Title" outline gradient size="small" onPress={cb} />);

      expect(console.warn).toHaveBeenCalledWith(
        "You can't use `outline` prop combined with `gradient` or `active` props.",
      );
    });

    it('logs children and title warn', () => {
      Mount(
        <Button text="Title" size="small" onPress={cb}>
          <></>
        </Button>,
      );

      expect(console.warn).toHaveBeenCalledWith(
        "You can't use `title` prop combined with `children`. Use only one of both.",
      );
    });
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(ThemeContext(<Button text="Title" size="small" rounded gradient onPress={cb} />))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
