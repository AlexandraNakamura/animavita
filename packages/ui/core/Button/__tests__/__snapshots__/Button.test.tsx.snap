// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  gradient={true}
  onPress={[MockFunction onPress]}
  rounded={true}
  size="small"
  style={
    Array [
      Object {
        "borderRadius": 20,
        "justifyContent": "center",
      },
    ]
  }
>
  <View
    colors={
      Array [
        "#1BC0B5",
        "#2ED68F",
      ]
    }
    end={
      Object {
        "x": 1,
        "y": 1,
      }
    }
    gradient={true}
    locations={
      Array [
        0.1,
        0.9,
      ]
    }
    rounded={true}
    size="small"
    start={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 20,
          "paddingBottom": 8,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 8,
        },
      ]
    }
  >
    <Text
      color="#FFFFFF"
      style={
        Array [
          Object {
            "color": "#353434",
          },
          Object {
            "color": "#FFFFFF",
            "fontSize": 26.5,
            "fontWeight": "bold",
          },
          Object {
            "textAlign": "center",
          },
        ]
      }
      themeName="light"
      type="bold"
    >
      Title
    </Text>
  </View>
</TouchableOpacity>
`;
